### DISCUSSION TITLE
Entity relationships now break with class-transformer

### DISCUSSION THREAD

**Discussion by noftaly** [mikro-orm/mikro-orm#2687] [Q&A]
Hi there!

**Describe the bug**
I'm using Nest.js with MikroORM and recently experienced a crash when upgrading to `v4.5.10` (which is not fixed in `v5.0.0-rc.0`).
Whenever I'm returning an entity with a relationship, while having class-transformer enabled, the app crashes.
The stack trace comes from class-transformer, although it worked fine before upgrading to v4.5.10 so I assume this comes from MikroORM. If not, sorry for the trouble, I'll open an issue on class-transformer's repo :)

**Stack trace**
```
LOG [MikroORM] [query] select "e0".* from "post" as "e0" [took 2 ms]
LOG [MikroORM] [query] select "e0".*, "e1"."user_user_id" as "fk__user_user_id", "e1"."post_post_id" as "fk__post_post_id" from "user" as "e0" left join "post_assignees" as "e1" on "e0"."user_id" = "e1"."user_user_id" where "e1"."post_post_id" in (1) [took 3 ms]
ERROR [ExceptionsHandler] Maximum call stack size exceeded
RangeError: Maximum call stack size exceeded
    at TransformOperationExecutor.transform (mikro-orm-repro/node_modules/src/TransformOperationExecutor.ts:117:29)
    at TransformOperationExecutor.transform (mikro-orm-repro/node_modules/src/TransformOperationExecutor.ts:327:31)
    at TransformOperationExecutor.transform (mikro-orm-repro/node_modules/src/TransformOperationExecutor.ts:327:31)
    at TransformOperationExecutor.transform (mikro-orm-repro/node_modules/src/TransformOperationExecutor.ts:327:31)
    at TransformOperationExecutor.transform (mikro-orm-repro/node_modules/src/TransformOperationExecutor.ts:327:31)
    at TransformOperationExecutor.transform (mikro-orm-repro/node_modules/src/TransformOperationExecutor.ts:327:31)
    at TransformOperationExecutor.transform (mikro-orm-repro/node_modules/src/TransformOperationExecutor.ts:327:31)
    at TransformOperationExecutor.transform (mikro-orm-repro/node_modules/src/TransformOperationExecutor.ts:327:31)
    at TransformOperationExecutor.transform (mikro-orm-repro/node_modules/src/TransformOperationExecutor.ts:327:31)
    at TransformOperationExecutor.transform (mikro-orm-repro/node_modules/src/TransformOperationExecutor.ts:327:31)
```

**To Reproduce**
Minimal reproduction repo : https://github.com/noftaly/mikro-orm-repro
Curl command : `curl --request POST --url http://localhost:3030/posts`
and `curl --request GET --url http://localhost:3030/posts`

**Expected behavior**
I expected it not to crash, as it used to do on v4.5.9.

**Versions**

| Dependency | Version |
| - | - |
| node | v16.13.0 |
| typescript | 4.3.5 |
| mikro-orm | >=4.5.10 |
| your-driver | PostgreSQL >=4.5.10 |

**Comment by B4nan** [comment]
As you said, this is not coming from MikroORM, so I doubt it is something fix in here. Note that you should _not_ use class-transformer on MikroORM entities, as it won't respect the constraints of such (e.g. reference and collection wrappers, populate hints, custom serializers). 

Serialization should be handled via `toJSON()/toObject()` method that is automatically defined on the entity prototype - this way serialization works based on how you load things (populate hints), this information is not available to class-transformer. This method is called automatically when serializing the response to JSON (so when calling `JSON.stringify(entity)`), so you can just return your entity from nest endpoints and it will be called for you anyway. You don't really need class-transformer.

Creating entities from DTOs need to be handled via `em.create()` or similar methods - that way it will create references and collections where they should be.

**Reply by noftaly** [reply]
Nice thank you! I'll use `toJSON()` from now on ;)

**Reply by noftaly** [reply]
I have another question now that I'm implementing this.
How can I implement groups? `class-transformer` allows us to choose when a property will be show, like this:
```js
// user.entity.ts
@Property({ type: 'text' })
@Expose({ groups: ['EMAIL_INCLUDED'] })
email: string;

// user.controller.ts
@SerializeOptions({ groups: ['EMAIL_INCLUDED'] })
@Get()
getUser(): User {} // email will be returned


@Get()
getUser2(): User {} // email won't be returned
```

In a simple `toJSON()` method which does not have access to the metadata, this does not seem possible. It looks like I'll have to recode most of class-transformer's decorators to achieve the same feature.

Are groups possible, have I missed something?
Otherwise, because MikroORM used to work with class-transformer, is it not possible to patch MikroORM to bring back this support? `class-transformer` really is at the center of Nest's ecosystem, so I think this would benefit many people 😄 

Thanks :)

**Reply by B4nan** [reply]
You can access the metadata anywhere, they are linked on the entity prototype, so `wrap(this, true).__meta` (or simple `(this as AnyEntity).__meta` could work too). You could even access it statically without entity instance, I guess something like `wrap(MyEntity.prototype, true).__meta` should work.

> Otherwise, because MikroORM used to work with class-transformer, is it not possible to patch MikroORM to bring back this support? class-transformer really is at the center of Nest's ecosystem, so I think this would benefit many people 😄

Well, as said in the first reply, this is not something to be addressed in MikroORM, if you see cycle from what the class-transofmer does, it is not our fault - there are many cycles in entity definition, especially in internal properties like `__meta`. They are not enumerable, so it should not matter. But there can be enumerable cycles too, e.g. if you have a self referencing relation.

And again, you can use it, you just can't feed it the entity instance - you need to call `toObject` or `toJSON` before you return it.

```ts
  @Post()
  async create() {
    const entity = await this.postsService.create();
    return wrap(entity).toJSON();
  }
```

I will try to review your repro when I have some spare time, maybe it is something we could fix on our end, but even if that would be the case - to me this is a class-transformer bug most probably.

**Reply by Jedliu** [reply]
I'm having the same problem both in class-transformer and [accesscontrol](https://github.com/onury/accesscontrol) filters. The filters will enumerate all the properties and the `Maximum call stack size exceeded` error will occur too when the relation is set.

**Reply by B4nan** [reply]
You cant work with entities with those tools as they dont understand how to work with them. Use the ORM to serialize (e.g. The toJSON method mentioned above).

**Reply by Jedliu** [reply]
I figure it out. First call toJson then do the filtering. It works for filtering only. I will skip the class-transformer. Thank you,
